lib/apimlStorage.ts

```diff
@@ -0,0 +1,382 @@
+/*
+  This program and the accompanying materials are
+  made available under the terms of the Eclipse Public License v2.0 which accompanies
+  this distribution, and is available at https://www.eclipse.org/legal/epl-v20.html
+  
+  SPDX-License-Identifier: EPL-2.0
+  
+  Copyright Contributors to the Zowe Project.
+*/
+
+import * as https from 'https';
+import * as http from 'http';
+import axios from 'axios';
+import { AxiosInstance } from 'axios';
+
+let apimlClient: AxiosInstance;
+
+export function configure(settings: ApimlStorageSettings) {
+  apimlClient = axios.create({
+    baseURL: `https://${settings.host}:${settings.port}`,
+    httpsAgent: new https.Agent(settings.tlsOptions)
+  });
+}
+
+export function isConfigured(): boolean {
+  return apimlClient != null;
+}
+
+export interface ApimlStorageSettings {
+  host: string;
+  port: number;
+  tlsOptions: https.AgentOptions;
+}
+
+
+export function makeStorageForPlugin(pluginId: string): ApimlStorage {
+  return new ApimlStorage(pluginId);
+}
+
+const CACHING_SERVICE_URI = '/cachingservice/api/v1/cache';
+
+interface ApimlRequest {
+  method: 'GET' | 'POST' | 'PUT' | 'DELETE',
+  path: string;
+  headers?: { [key: string]: string };
+  body?: any;
+}
+
+interface KeyValuePair {
+  key: string;
+  value: any;
+}
+interface ApimlResponse {
+  statusCode: number;
+  json?: any;
+  headers: http.IncomingHttpHeaders;
+}
+
+export type ApimlStorageErrorCode =
+  'APIML_STORAGE_CONNECTION_ERROR' |
+  'APIML_STORAGE_KEY_NOT_FOUND' |
+  'APIML_STORAGE_JSON_ERROR' |
+  'APIML_STORAGE_UNKNOWN_CLIENT_CERT' |
+  'APIML_STORAGE_UNAUTHORIZED' |
+  'APIML_STORAGE_NO_AUTH_PROVIDED' |
+  'APIML_STORAGE_INVALID_PAYLOAD' |
+  'APIML_STORAGE_UNKNOWN_ERROR' |
+  'APIML_STORAGE_RESPONSE_ERROR' |
+  'APIML_STORAGE_NOT_CONFIGURED';
+
+export class ApimlStorageError extends Error {
+  constructor(
+    public readonly code: ApimlStorageErrorCode,
+    private cause?: Error,
+    private apimlResponse?: ApimlResponse) {
+    super(code);
+    Object.setPrototypeOf(this, ApimlStorageError.prototype);
```

Comment:
Is this needed? I thought typescript does this for you.

Reply:
Yes, it is needed because of https://github.com/Microsoft/TypeScript-wiki/blob/master/Breaking-Changes.md#extending-built-ins-like-error-array-and-map-may-no-longer-work